### PR TITLE
[codex] Preserve single eForm PDF in fax preview

### PR DESCRIPTION
**Pull request by CecilFoubert**
## Summary
- switch eForm PDF generation to a browser-rendered path that preserves the rendered form surface
- allow local renderer requests to reach `efmshowform_data` without the HTML/script wrappers that were corrupting output
- preserve the original browser-rendered eForm PDF in fax preview when there are no attached documents, instead of rewriting it through `concatPDF`

## Root cause
Fax preview was not consuming a faithful eForm rendering. The browser-rendered path needed a local renderer endpoint, authenticated access to the form surface, direct image streaming for local assets, and wrapper suppression so the PDF saw the actual form markup. After that, the no-attachment fax case was still being rewritten through `DocumentAttachmentManagerImpl.concatPDF(...)`, which unnecessarily replaced the original renderer output.

## Impact
- single-document eForm fax preview now keeps the original `eform-browser-render-*.pdf` output instead of routing it through `combinedPDF_*`
- browser-rendered eForm PDFs retain the rendered form surface and local image/background access more reliably
- follow-up for the remaining multi-document concat/rewrite path is tracked in #3163

## Verification
- `JAVA_TOOL_OPTIONS='-Djava.awt.headless=true' mvn -Dtest=io.github.carlos_emr.carlos.documentManager.DocumentAttachmentManagerImplUnitTest,io.github.carlos_emr.carlos.eform.util.EFormBrowserPdfRendererUnitTest,io.github.carlos_emr.carlos.web.eform.EformViewForPdfGenerationServletUnitTest,io.github.carlos_emr.carlos.eform.util.EFormImageViewForPdfGenerationServletTest,io.github.carlos_emr.carlos.eform.util.EFormViewForPdfGenerationServletUnitTest,io.github.carlos_emr.carlos.managers.EformDataManagerImplCreatePdfUnitTest test`
  - `Tests run: 24, Failures: 0, Errors: 0, Skipped: 0`
- runtime verification in the clean `develop` worktree showed fax preview now requesting `faxFilePath=%2Fusr%2Flocal%2Ftomcat%2Ftemp%2Feform-browser-render-...pdf` with no browser-side console or page errors

---
## Summary by cubic
Keeps the original browser-rendered eForm PDF in fax preview when it’s the only document and strengthens the browser PDF rendering/preview path with better auth, validation, temp handling, and diagnostics. Fax preview now shows server-rendered page images with an “Open PDF” link.

- **New Features**
  - Added `EFormBrowserPdfRenderer` using local `playwright` with secure launch, hidden helper UI, temp isolation, strict local/private BASE_URL checks, path validation, session‑provider forwarding, and pre‑provisioned output dirs; integrated with `EformDataManagerImpl`.
  - Restored PDF assets and routes (fallback `signature_pad.min.js`, `jquery-3.7.1.min.js`, legacy `signatureControl.jsp` alias) and added local Playwright regression checks for key render surfaces.

- **Bug Fixes**
  - Preserve the original `eform-browser-render-*.pdf` when it’s the only document (skip concat); promote renderer/upload temp files into permanent storage before faxing.
  - Set correct content types and inline PDF `Content-Disposition`; return 400 for image requests without an extension.
  - Restore auth on the pdf-ready route (removed from unauthenticated list) and enforce `_fax` privilege on preview; skip CSRF/Logout HTML injection when marked by the renderer, including when set during the chain.
  - Restrict preview cache inputs to allowed temp/doc roots with safer validation and cleanup; improve renderer failure diagnostics with structured logs and redacted URLs while avoiding provider values or output content; strengthen error handling in eForm image streaming.
  - Addressed SonarCloud findings across the renderer and preview paths with stricter URL/host checks, safer logging, and more reliable temp/dir handling.

<sup>Written for commit 3f9d6986d3989ecc4e19148b854d605ed69cdc31. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/carlos-emr/carlos/pull/3164?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/main/java/io/github/carlos_emr/carlos/eform/util/EFormBrowserPdfRenderer.java [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+            deleteRecursivelyQuietly(outputDirectory);
+            deleteQuietly(outputPdfPath);
+            throw new PDFGenerationException("Unable to start the browser PDF renderer for eForms. " + abbreviate(processOutput), e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            deleteRecursivelyQuietly(outputDirectory);
+            deleteQuietly(outputPdfPath);
+            throw new PDFGenerationException("Browser rendering was interrupted while generating the eForm PDF.", e);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (CecilFoubert) [reply src/main/java/io/github/carlos_emr/carlos/eform/util/EFormBrowserPdfRenderer.java:201]: Current renderer path uses `--output-dir` intentionally: the Playwright script captures `page-*.png` files and the Java side assembles them into the final PDF. That flow is covered by the current renderer script and verified in the latest fax-preview runs, so this mismatch report is no longer applicable.

File: src/main/java/io/github/carlos_emr/carlos/eform/util/EFormBrowserPdfRenderer.java [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+        command.add("--app-path");
+        command.add(appPath);
+        command.add("--output-dir");
+        command.add(outputDirectory.toAbsolutePath().toString());
+        if (chromePath != null && !chromePath.isBlank()) {
+            command.add("--chrome-path");
+            command.add(chromePath);
+        }
+        return command;
+    }
```
[the remark this answers is not in the document]
Comment (CecilFoubert) [reply src/main/java/io/github/carlos_emr/carlos/eform/util/EFormBrowserPdfRenderer.java]: The current renderer contract is `--output-dir`, not `--output-path`: the Node renderer accepts `--output-dir`, writes page captures into that directory, and the Java renderer converts those captures into the final PDF. This review point is outdated against the current implementation.

File: src/main/java/io/github/carlos_emr/carlos/eform/util/EFormBrowserPdfRenderer.java [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+                }
+            }
+            document.save(outputPdfPath.toFile());
+        } catch (IOException e) {
+            throw new PDFGenerationException("Unable to assemble the browser-rendered eForm captures into a PDF.", e);
+        }
+    }
+
+    private static boolean isDefaultPort(String scheme, int port) {
+        return ("http".equalsIgnoreCase(scheme) && port == 80)
```
[the remark this answers is not in the document]
Comment (CecilFoubert) [reply src/main/java/io/github/carlos_emr/carlos/eform/util/EFormBrowserPdfRenderer.java:667]: The current static-analysis run for this revision passes `SpotBugs + Find Security Bugs`, and this `IMPROPER_UNICODE` finding is no longer present in the live issue set for PR 3164. This thread is stale against the current renderer code.

File: src/main/java/io/github/carlos_emr/carlos/eform/util/EFormBrowserPdfRenderer.java [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+    }
+
+    static String validateRendererBaseUrl(String rawBaseUrl) {
+        if (rawBaseUrl == null || rawBaseUrl.isBlank()) {
+            throw new IllegalArgumentException("Renderer base URL must be non-empty");
+        }
+
+        URI uri = URI.create(rawBaseUrl.trim());
+        String scheme = uri.getScheme();
+        if (scheme == null || (!"http".equalsIgnoreCase(scheme) && !"https".equalsIgnoreCase(scheme))) {
```
[the remark this answers is not in the document]
Comment (CecilFoubert) [reply src/main/java/io/github/carlos_emr/carlos/eform/util/EFormBrowserPdfRenderer.java:353]: The current static-analysis run for this revision passes `SpotBugs + Find Security Bugs`, and this `IMPROPER_UNICODE` finding is no longer present in the live issue set for PR 3164. This thread is stale against the current renderer code.

File: src/main/java/io/github/carlos_emr/carlos/eform/util/EFormBrowserPdfRenderer.java [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+            }
+            document.save(outputPdfPath.toFile());
+        } catch (IOException e) {
+            throw new PDFGenerationException("Unable to assemble the browser-rendered eForm captures into a PDF.", e);
+        }
+    }
+
+    private static boolean isDefaultPort(String scheme, int port) {
+        return ("http".equalsIgnoreCase(scheme) && port == 80)
+                || ("https".equalsIgnoreCase(scheme) && port == 443);
```
[the remark this answers is not in the document]
Comment (CecilFoubert) [reply src/main/java/io/github/carlos_emr/carlos/eform/util/EFormBrowserPdfRenderer.java:668]: The current static-analysis run for this revision passes `SpotBugs + Find Security Bugs`, and this `IMPROPER_UNICODE` finding is no longer present in the live issue set for PR 3164. This thread is stale against the current renderer code.

**Comment by CecilFoubert** [comment]
@coderabbitai review

**Comment by yingbull** [comment]
@copilot+gpt-5.5 check all comments and resolve if valid but ensure we do not break pdf rendering again, the point of this pr.  respond to all comments you address, or determine are already fixed or incorrect and no action required.

**Comment by yingbull** [comment]
@copilot+claude-fable-5 I am seeing still a lot of open comments from cubic and other reviewers.  Run the check again.  make sure you comment on all open tickets even if they are false or already done or not needed.   for any you are going to fix, review carefully to ensure valid and still needed with current code.  It is critical your fixes do not break pdf rendering / preview that is the whole point of this pr.

File: src/main/java/io/github/carlos_emr/carlos/eform/util/EFormBrowserPdfRenderer.java [hunk 6]
```diff
@@ -0,0 +1,594 @@
...
+    @SuppressFBWarnings(value = "PATH_TRAVERSAL_IN", justification = "Renderer temp files are created only beneath resolveRendererTempRoot(), which validates configured roots before creating a managed private temp directory.")
+    private static Path createSecureTempPath(Path tempRoot, boolean directory, String prefix, String suffix) throws IOException {
+        Path managedRoot = Files.createDirectories(tempRoot);
+        FileAttribute<?>[] secureAttributes = securePosixAttributes(directory);
+        if (secureAttributes.length > 0) {
+            try {
+                return directory
+                        ? Files.createTempDirectory(managedRoot, prefix, secureAttributes)
+                        : Files.createTempFile(managedRoot, prefix, suffix, secureAttributes);
+            } catch (UnsupportedOperationException e) {
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/eform/util/EFormBrowserPdfRenderer.java]: seems odd, don’t think we need to make directories

**Comment by yingbull** [comment]
@copilot+claude-opus-4.8 like the last follow up on open review comments.

**Comment by CecilFoubert** [comment]
SonarCloud on current head `30ece19498` is down to 7 open issues, all of which are being left intentionally for this PR.

Rationale by bucket:
- `java:S6541` brain-method findings in `NioFileManagerImpl`, `Fax2Action`, `EFormPDFServlet`, and `EForm` are maintainability-only refactors, not correctness or security issues. Clearing them would require broader method reshaping in renderer/fax/eForm code paths that are already behaving correctly and are the primary risk area for this PR.
- `java:S135` in `EFormViewForPdfGenerationServlet` is a low-value loop-style complaint (`continue` count) with no effect on behavior, security, or merge readiness.
- `java:S1075` in `EFormAssetDeployer` is the generic hardcoded-path/URI-style rule on an internal WAR resource path constant, not an unsafe external path or user-controlled filesystem boundary.

At this point the remaining Sonar findings are code-smell noise or larger cleanup work, while the PR’s functional/security findings and renderer regressions have been addressed. Given the explicit requirement not to destabilize PDF rendering / preview, I am intentionally stopping here rather than churning those code paths further in this branch.

**Comment by Ben-Heerema** [comment]
Hey @CecilFoubert, 

I think this PR currently breaks saved eForm PDF download.

Evidence:
  - `develop` SHA `65bb859076748b2df811c3ebc9a9671338219fd9`
  - Same flow produced a valid PDF
  - PDF size: `4,849` bytes
  - Header: `%PDF-`

  - PR SHA `30ece1949880d7b4d23d07ef075102273755a6ea`
  - Same flow timed out waiting for a browser download
  - UI showed: `This eForm (and attachments, if applicable) could not be downloaded.`
  - No PDF artifact was produced
  - Server logged: `PDFGenerationException: Browser rendering failed while generating the eForm PDF. exitStatus=1`

The likely issue is that the new browser renderer builds this URL:

`/eformViewForPdfGenerationServlet?fdid=...`

But in the live app, that lowercase route returned an empty page for me. The uppercase PDF servlet route rendered the actual eForm content:

`/EFormViewForPdfGenerationServlet?fdid=...&browserRender=true`

So the renderer is effectively trying to screenshot a blank page, then failing.

I’d recommend updating the renderer to use the actual PDF-ready servlet route and adding an integration/browser test that verifies the renderer sees non-empty eForm content and produces a PDF. Also worth logging the Node renderer stderr/stdout, since the Java exception currently only shows `exitStatus=1`.

**Comment by phc007** [comment]
@CecilFoubert 
I'm getting an error on building this branch after this AM's pull git switch

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.15.0:compile (default-compile) on project carlos: Compilation failure: Compilation failure: 
[ERROR] /home/peter/carlos/src/main/java/io/github/carlos_emr/carlos/encounter/oscarConsultationRequest/pageUtil/ConsultationPDFCreator.java:[816,9] cannot find symbol
[ERROR]   symbol:   variable table
[ERROR]   location: class io.github.carlos_emr.carlos.encounter.oscarConsultationRequest.pageUtil.ConsultationPDFCreator
[ERROR] /home/peter/carlos/src/main/java/io/github/carlos_emr/carlos/encounter/oscarConsultationRequest/pageUtil/ConsultationPDFCreator.java:[817,16] cannot find symbol
[ERROR]   symbol:   variable table
[ERROR]   location: class io.github.carlos_emr.carlos.encounter.oscarConsultationRequest.pageUtil.ConsultationPDFCreator
[ERROR] -> [Help 1]

**Comment by CecilFoubert** [comment]
> @CecilFoubert I'm getting an error on building this branch after this AM's pull git switch
> 
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.15.0:compile (default-compile) on project carlos: Compilation failure: Compilation failure: [ERROR] /home/peter/carlos/src/main/java/io/github/carlos_emr/carlos/encounter/oscarConsultationRequest/pageUtil/ConsultationPDFCreator.java:[816,9] cannot find symbol [ERROR] symbol: variable table [ERROR] location: class io.github.carlos_emr.carlos.encounter.oscarConsultationRequest.pageUtil.ConsultationPDFCreator [ERROR] /home/peter/carlos/src/main/java/io/github/carlos_emr/carlos/encounter/oscarConsultationRequest/pageUtil/ConsultationPDFCreator.java:[817,16] cannot find symbol [ERROR] symbol: variable table [ERROR] location: class io.github.carlos_emr.carlos.encounter.oscarConsultationRequest.pageUtil.ConsultationPDFCreator [ERROR] -> [Help 1]

OK, I'll look into it

**Comment by CecilFoubert** [comment]
OK, @phc007 both Ben and I had no luck re-creating your issue, so we think it might be a local problem

**Comment by phc007** [comment]
@CecilFoubert 
thanks for checking, 
deleting the directory and getting a fresh clone fixed it!

**Comment by phc007** [comment]
@CecilFoubert
testing on production data the current state of build

1) faxing an eform gives me
2026-07-14 10:53:55] [info] 2026-07-14 10:53:55,957 ERROR action.Fax2Action (Fax2Action.java:402) - EForm PDF generation failed during browser rendering.
[2026-07-14 10:53:55] [info] io.github.carlos_emr.carlos.utility.PDFGenerationException: EForm PDF generation failed during browser rendering.
[2026-07-14 10:53:55] [info] #011at io.github.carlos_emr.carlos.managers.EformDataManagerImpl.createEformPDF(EformDataManagerImpl.java:200)
[2026-07-14 10:53:55] [info] #011at io.github.carlos_emr.carlos.documentManager.DocumentAttachmentManagerImpl.renderEFormWithAttachments(DocumentAttachmentManagerImpl.java:509)
[2026-07-14 10:53:55] [info] #011at io.github.carlos_emr.carlos.fax.action.Fax2Action.prepareFax(Fax2Action.java:395)
.....
[2026-07-14 10:53:55] [info] Caused by: java.lang.NullPointerException: Cannot invoke "java.nio.file.Path.getFileSystem()" because "path" is null

2) Attempting to preview an attachment in an consult request yields the warning
A preview of this document could not be generated.
Failed to render eForm PDF.

[2026-07-14 10:57:31] [info] 2026-07-14 10:57:31,006 ERROR actions.DocumentPreview2Action (DocumentPreview2Action.java:204) - Error occurred while rendering eForm. EForm PDF generation failed during browser rendering.
[2026-07-14 10:57:31] [info] io.github.carlos_emr.carlos.utility.PDFGenerationException: EForm PDF generation failed during browser rendering.
[2026-07-14 10:57:31] [info] #011at io.github.carlos_emr.carlos.managers.EformDataManagerImpl.createEformPDF(EformDataManagerImpl.java:200)

**Comment by CecilFoubert** [comment]
Yeah, I know, trying to fix it all today. Thanks for the error messages though, those can be pretty helpful!

**Comment by phc007** [comment]
@CecilFoubert
testing on production data the current state of build


    Attempting to preview an attachment in an consult request yields the warning
    A preview of this document could not be generated.
    Failed to render eForm PDF.

without a catalina.out


    faxing an eform gives me a blank window
10.69.241.139:8443/carlos/fax/faxAction?method=prepareFax&transactionId=273058&transactionType=EFORM&demographicNo=8217
    
without a catalina.out

**Comment by Ben-Heerema** [comment]
@coderabbitai review
@cubic-dev-ai review

**Comment by CecilFoubert** [comment]
Confirming DCO sign off for all commits

File: src/main/java/io/github/carlos_emr/carlos/eform/util/EFormImageViewForPdfGenerationServlet.java [hunk 7]
```diff
@@ -13,48 +13,189 @@
...
 /**
  * The purpose of this servlet is to allow a local process to access eform images.
  */
 public final class EFormImageViewForPdfGenerationServlet extends HttpServlet {
 
     private static final Logger logger = MiscUtils.getLogger();
+    private static final String VACCINE_BRANDS_FILE = "vaccine-brands.json";
+    private static final String IMAGE_JPEG = "image/jpeg";
+    private static final String HTML_CONTENT_TYPE = "text/html";
+    private static final Map<String, String> CONTENT_TYPES = Map.ofEntries(
```
[the remark this answers is not in the document]
Comment (CecilFoubert) [reply src/main/java/io/github/carlos_emr/carlos/eform/util/EFormImageViewForPdfGenerationServlet.java:48]: Leaving this open for now. I still think the duplication concern is real in the abstract, but I do not want this PR to couple the hardened renderer-only asset servlet to the broader `/eform/displayImage` response path. The current servlet intentionally serves a very small allow-list for loopback PDF capture, and extracting shared metadata/helpers here risks widening that surface or reintroducing behavior differences through a shared abstraction. If we want to deduplicate it, that should be a follow-up refactor with its own tests around supported asset types and response headers rather than something folded into the fax/eForm rendering fix set.

**Comment by phc007** [comment]
@CecilFoubert
testing on production data the current state of build https://github.com/carlos-emr/carlos/commit/aa57cdb009869df47587c4f1af8a60dfd8fbe6a0
on Firefox

Attempting to preview an attachment in an consult request yields the warning
A preview of this document could not be generated.
Failed to render eForm PDF.

without a catalina.out or a console.log

faxing an eform gives me a blank window
10.69.241.139:8443/carlos/fax/faxAction?method=prepareFax&transactionId=273060&transactionType=EFORM&demographicNo=8217

without a catalina.out or a console.log

**Comment by CecilFoubert** [comment]
Alright @phc007 this patch should give better logging at least, might help figure out exactly what's breaking

**Comment by phc007** [comment]
@CecilFoubert
testing on production data this morning with a fresh pull
on Firefox

Attempting to preview an attachment in an consult request yields the warning popup
A preview of this document could not be generated.
Failed to render eForm PDF.

without a catalina.out or a console.log

faxing an eform gives me a blank window
10.69.241.139:8443/carlos/fax/faxAction?method=prepareFax&transactionId=273060&transactionType=EFORM&demographicNo=8217

without a catalina.out or a console.log

**Comment by CecilFoubert** [comment]
> @CecilFoubert testing on production data this morning with a fresh pull on Firefox
> 
> Attempting to preview an attachment in an consult request yields the warning popup A preview of this document could not be generated. Failed to render eForm PDF.
> 
> without a catalina.out or a console.log
> 
> faxing an eform gives me a blank window 10.69.241.139:8443/carlos/fax/faxAction?method=prepareFax&transactionId=273060&transactionType=EFORM&demographicNo=8217
> 
> without a catalina.out or a console.log

Huh, that's odd, I'll look into it more, one of the possible causes that popped up on my end was that it might be rejecting based on host (i.e. your IP port combo), so I'll do some more testing and digging for a cause.

**Comment by CecilFoubert** [comment]
Confirming DCO sign off for all commits

**Comment by CecilFoubert** [comment]
Remaining SonarQube issues are just cleanliness really, so I think this PR can move forwards, once we get things figured out with Peter's errors

**Comment by phc007** [comment]
@CecilFoubert 
don't merge this yet as now I am not even getting the pdf from the consult request when it has an eform attachment

To start neither Blink nor Gecko resolves a preview of an eform for attachment for me
even after a fresh clone and git switch to the branch to ensure I'm not compiling something stale


you need to add some debugging code as nothing now is throwing in js console nor catalina which is going to make this hard for AI or human oversight to track down

**Comment by CecilFoubert** [comment]
Closing as superseded: this branch's entire subject area (browser-rendered eForm PDF generation, fax preview handling) was superseded by the already-merged PR #3193 ("feat(eform): add secure Chromium text-layer PDF rendering and harden eForms, attachments, fax, and clinical document workflows"), which landed 2026-07-28.

Attempting to merge this now produces conflicts in 25+ files, including a modify/delete conflict on `EFormViewForPdfGenerationServlet.java` — that file was deleted on `develop` as part of #3193's rework, but this branch still modifies it. Reconciling this branch would mean a from-scratch reimplementation against the current eForm-rendering architecture, not a rebase.

If the original single-document fax-preview goal ("preserve the exact rendered PDF instead of re-rendering through concatPDF") is still not covered by #3193 + the newer PR #3233 ("Reuse staged eForm PDF for fax approval"), that's worth re-scoping as a fresh PR against current `develop` rather than reviving this one.
